File: 03 DYNAMO RESOURCES/Akaba-build-production/Akaba-build-production/services/generators/stuffer/Stuffer/Generator/Strategy/Phenotype/CreateSpaceStrategy.cpp
#include <stdafx.h>
#include <CreateSpaceStrategy.h>
#include <Args.h>
#include <BuildData.h>
#include <Phenotype.h>

CreateSpaceStrategy::CreateSpaceStrategy(int floor, const Rect2i& rect, Usage::name_tag tag)
: floor(floor),
  rect(rect),
  tag(tag)
{
}

const string_t& CreateSpaceStrategy::name() const
{
  static string_t name(U("CreateSpaceStrategy"));
  return name;
}

void CreateSpaceStrategy::execute(const Args& args) const
{
  auto& phenotype(args.getPhenotype());
  phenotype.createFloor(floor);
  phenotype.createSection(floor, rect, tag);
  args.getBuildData().setClaimed(floor, rect);
}

File: 03 DYNAMO RESOURCES/Akaba-build-production/Akaba-build-production/services/generators/hoover/stuffer/Strategy/CreateSpaceStrategy.cs
namespace stuffer
{
  public class CreateSpaceStrategy : Strategy
  {
    private int m_level;
    private Rect2i m_rect;
    private string m_usage;

    private CreateSpaceStrategy(StrategyArgs strategyArgs, int level, Rect2i rect, string usage)
    : base(strategyArgs)
    {
      m_level = level;
      m_rect = rect;
      m_usage = usage;
    }

    public static CreateSpaceStrategy Create(StrategyArgs strategyArgs, int level, Rect2i rect, string usage)
    {
      return new CreateSpaceStrategy(strategyArgs, level, rect, usage);
    }

    public override StrategyArgs Execute()
    {
      Building design = m_strategyArgs.Design;
      design.CreateFloor(m_level);
      design.CreateRegion(m_level, m_rect, m_usage);

      // TODO: Add claimed flag to grid after implementing
      //strategyArgs.getBuildData().setClaimed(level, rect);

      return m_strategyArgs;
    }

  }
}
